### PR TITLE
Added hid-sys library

### PR DESCRIPTION
**Pull request by Jonesey13**

### CODE REVIEW

File: src/hid.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
++// Copyright � 2015, Alex Daniel Jones
++// Licensed under the MIT License <LICENSE.md>
++//! Taken from hidclass.h, hidsdi.h and hidpi.h
```
Comment (retep998) [inline src/hid.rs]: Each header needs to have a corresponding module. So `hidclass.rs` `hidsdi.rs` and `hidpi.h`

File: src/hid.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    pub VendorID: ::USHORT,
+    pub ProductID: ::USHORT,
+    pub VersionNumber: ::USHORT,
+}
+pub type PHID_COLLECTION_INFORMATION = *mut HID_COLLECTION_INFORMATION;
+
+#[repr(C)] #[derive(Clone, Copy, Debug)]
+pub struct HIDP_UNKNOWN_TOKEN {
+    pub Token: ::UCHAR,
+    pub Reserved: [::UCHAR;3],
```
Comment (retep998) [inline src/hid.rs]: Missing space.

File: src/hid.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    pub BitField: ::ULONG,
+}
+pub type PHIDP_UNKNOWN_TOKEN = *mut HIDP_UNKNOWN_TOKEN;
+
+#[repr(C)] #[derive(Clone, Copy, Debug)]
+pub struct HIDP_EXTENDED_ATTRIBUTES {
+    pub NumGlobalUnknowns: ::UCHAR,
+    pub Reserved: [::UCHAR;3],
```
Comment (retep998) [inline src/hid.rs]: Missing space.

File: src/hid.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    pub GlobalUnknowns: ::PHIDP_UNKNOWN_TOKEN,
+    pub Data: [::ULONG;1],
```
Comment (retep998) [inline src/hid.rs]: Missing space.

File: src/hid.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    pub FirstChild: ::USHORT,
+    pub CollectionType: ::ULONG,
+    pub IsAlias: ::ULONG,
+    pub Reserved: ::ULONG,
+    pub UserContext: ::PVOID,
+}
+pub type PHIDP_LINK_COLLECTION_NODE = *mut HIDP_LINK_COLLECTION_NODE;
+
+#[repr(C)] #[derive(Clone, Copy, Debug)]
+pub struct HIDP_CAPS
```
Comment (retep998) [inline src/hid.rs]: Keep opening braces on the same line please.

File: src/hid.rs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    pub IsAbsolute: ::BOOLEAN,
+
+    pub Reserved: [::ULONG; 10],
+    pub Range: RANGE_STRUCT,
+}
+
+pub type PHIDP_BUTTON_CAPS = *mut HIDP_BUTTON_CAPS;
+
+#[repr(C)] #[derive(Clone, Copy, Debug)]
+pub struct RANGE_STRUCT{
```
Comment (retep998) [inline src/hid.rs]: Missing space.

File: src/hid.rs (hunk 1 above)
Comment (Jonesey13) [inline src/hid.rs]: Ah I was wondering about that; Didn't want to clutter up the src folder with multiple files. Will correct.

File: lib/hid-sys/src/lib.rs [hunk 7]
```diff
@@ -4,49 +4,194 @@
...
 use winapi::*;
+
 extern "system" {
-    // pub fn HidD_FlushQueue();
-    // pub fn HidD_FreePreparsedData();
-    // pub fn HidD_GetAttributes();
+    pub fn HidP_GetCaps(
+        PreparsedData: PHIDP_PREPARSED_DATA, Capabilities: PHIDP_CAPS,
+    ) -> NTSTATUS;
+
```
Comment (retep998) [inline lib/hid-sys/src/lib.rs]: Please no blank lines between each function. It wastes vertical space.

File: lib/hid-sys/src/lib.rs [hunk 8]
```diff
@@ -4,49 +4,194 @@
...
+        ReportType: HIDP_REPORT_TYPE, UsagePage: USAGE, LinkCollection: USHORT,
+        UsageList: PUSAGE, UsageLength: PULONG, PreparsedData: PHIDP_PREPARSED_DATA,
+        Report: PCHAR, ReportLength: ULONG,
+    ) -> NTSTATUS;
+
+    pub fn HidP_UsageListDifference(
+        PreviousUsageList: PUSAGE, CurrentUsageList: PUSAGE, BreakUsageList: PUSAGE,
+        MakeUsageList: PUSAGE, UsageListLength: ULONG,
+    ) -> NTSTATUS;
+
```
Comment (retep998) [inline lib/hid-sys/src/lib.rs]: Could you keep all the functions sorted by name?

**Comment by retep998** [comment]
Make sure all the files are encoded in UTF-8.

File: src/hidpi.rs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    pub DataIndexMax: ::USHORT,
+}
+
+#[repr(C)] #[derive(Clone, Copy, Debug)]
+pub struct HIDP_VALUE_CAPS {
+    pub UsagePage: ::USAGE,
+    pub ReportID: ::UCHAR,
+    pub IsAlias: ::BOOLEAN,
+
+    pub BitField: ::USHORT,
```
Comment (retep998) [inline src/hidpi.rs]: If this really is a bitfield, you can take advantage of the `BITFIELD!` macro to allow access to those fields.

Comment (Jonesey13) [inline src/hidpi.rs]: In this particular case it looks like it's just a variable name: https://msdn.microsoft.com/en-us/library/windows/hardware/ff539832(v=vs.85).aspx however HIDP_LINK_COLLECTION_NODE  has bitfields and has to be fixed. Also I should add some UNION! macros here (very handy). Wished I knew about this macro when I was using RAWINPUT.

**Pull request merged**